agent/os_helper.py

```diff
@@ -303,3 +316,52 @@ def apply(self, pkg):
                 'source_name': kernel_pkg[rpm.RPMTAG_NAME].decode(),
                 'source_version': kernel_pkg[rpm.RPMTAG_EVR].decode()
             }
+
+
+def get_latest_same_kernel_deb(name_part0, name_part2):
+    """
+    Return the latest version of a deb package for given name parts.
+    """
+    import apt
+    search_pattern = re.compile(name_part0 + r'(\d+)' + name_part2)
+
+    class KernelFilter(apt.cache.Filter):
+        """Filter class for checking for matching with a RE search pattern."""
+        def apply(self, pkg):
+            return pkg.is_installed and search_pattern.match(pkg.name)
+
+    cache = apt.cache.FilteredCache(apt.Cache())
+    cache.set_filter(KernelFilter())
+    return sorted([(int(search_pattern.match(deb.name).group(1)), deb) for deb in cache],
+                  reverse=True)[0][1]
+
+
+def reboot_required():
+    """
+    Check if reboot required by comparing running kernel package's version
+     with the newest installed kernel package's one.
+    """
+    boot_image_path = kernel_cmdline().get('BOOT_IMAGE')
+    if boot_image_path is None:
+        return None
+    if is_debian():  # For apt-based distros.
+        import apt_pkg
+        apt_pkg.init()
+        kernel_pkg = get_kernel_deb_package(boot_image_path)
+        if kernel_pkg is not None:
+            match = DEBIAN_KERNEL_PKG_NAME_RE.match(kernel_pkg.name)
+            if match:
+                name_parts = match.groups()  # E.g. ('linux-image-4.4.0-', '174', '-generic')
```

Comment:
How about we change this regex to `linux-image-\d+\.\d+\.\d+\.-\d+-.+` since we actually expect it to look like that? This way we don't accidentally catch packets which we don't know.